Initial Dockerfile for codespaces...

coatless opened the pull request:
I've tried to keep this `Dockerfile` as minimal as possible. The target for the container is to quickly bootstrap contributing to `mlpack` via [GitHub Codespaces](https://github.com/features/codespaces).

Unlike the `mlpack/mlpack` and other containers in the repository, we're building this off of the [universal devcontainer](https://github.com/devcontainers/images/tree/main/src/universal) to lower the storage requirements associated with deployment on GitHub Codespaces following the ["Getting the most out of your included usage"](https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-included-usage#about-codespaces-storage).

coatless committed:
Initial Dockerfile for codespaces...

coatless commented:
Welp, this isn't minimal at all in terms of the base image. Building onto the container with just a few items causes the image to exceed what's available on GH Action Builder to create a modified image. Plus, I can't build the image locally due to it rejecting `arm64` 😢 

So, my thought here is to move away from the universal as a base image or lobotomize the container of `.NET`/`Java`/`PHP`/ ... :

https://github.com/devcontainers/images/blob/main/src/universal/history/2.9.0.md

There are other containers, but we're stuck with a container that uses the Universal ID if we want it to work 👎 . 

Any preference @rcurtin ?

rcurtin commented:
Do you think I could build it locally and push it?  It's ok with me if each container needs to be handbuilt and pushed to Dockerhub to be updated---that's already what we do for containers when mlpack versions get bumped.

Happy to try it.  I do have an OS X arm64 (M1 or M2, can't remember) system I could use to build that version.

coatless commented:
So, the hiccup building locally was the base universal is `linux/amd64` and doesn't play nicely with M-series macs (`linux/arm64`). Though, under `buildx` build emulation, I snagged a viable [image now on DockerHub](https://hub.docker.com/r/jamesbalamuta/mlpack-codespace/tags). 

```sh
docker buildx build --platform=linux/amd64 -t mlpack-codespace .
```

I'll have time to check the image later today that this `Dockerfile` generated.

rcurtin commented:
Awesome!  If that works, I can push it to the mlpack Dockerhub account too.

rcurtin commented on Dockerfiles/mlpack-codespace/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+    libblas-dev \
+    libarpack2 \
+    libsuperlu-dev \
+    libstb-dev && \
```
I think we will also need `libcereal-dev`, unless that's already installed. :+1:

coatless replied:
Added

rcurtin commented on Dockerfiles/mlpack-codespace/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+    pandas setuptools
+
+## Install armadillo.
+RUN curl -Lk https://files.mlpack.org/armadillo-11.4.1.tar.gz | tar -xvz && \
```
Happy to put a newer version in place if you like.

coatless replied:
We'd want to bump the version associated with all images if that's the case. I think conrad just bumped the current default to: 10.8.2. So, we're potentially running one version higher than we should...

rcurtin replied:
That's only the minimum version, so I don't think it's a problem to have a newer one.  But if you want to keep it at the minimum, I put https://files.mlpack.org/armadillo-10.8.2.tar.gz in place.

coatless replied:
We probably should make sure everything is running on the minimums to avoid running into a hidden bug (even though the API is pretty stable).

coatless replied:
Switched it over to the minimum version.

rcurtin commented:
@coatless just a ping on this one---let me know if I can be helpful in keeping things moving here. :+1:

coatless committed:
Merge branch 'mlpack:master' into universal-dev-container-for-codespaces

coatless committed:
Add libcereal-dev

coatless committed:
Prevent macOS system files from being included in the input.

coatless committed:
Explore adding a step on R

coatless commented:
@rcurtin at the moment, the slow down is more or less an internet cap by lovely Comcast. I can only really push docker builds and validate at end of month or if I know I'm going to be on a stable fast internet connection elsewhere.

coatless committed:
Clean up the R install step

Build R based on the official instructions
Retrieve required packages for development through r2u

coatless committed:
Handle docker linting warnings:

1. LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 6, 68)

And

2. JSONArgsRecommended: JSON arguments recommended for CMD to prevent unintended behavior related to OS signals (line 69)

rcurtin commented:
Ha, could you borrow Red Rock's wireless? 😄 I can provide a system you can ssh to that has docker on it and no bandwidth cap, so if that would help fix the issue for you (or rather just work around the awful that is Comcast...) just let me know.

coatless commented:
@rcurtin coffee shop wireless with a 10gb image is _painfully_ slow. So, if you have a spare box that I can SSH into for this, I'd be game. 

The next steps in this process:

1. Check that everyone is happy with the above `Dockerfile`
2. Place the image underneath MLPACK's docker hub: <https://hub.docker.com/u/mlpack>
3. Document in the main repo: https://github.com/mlpack/mlpack/pull/3637

coatless committed:
Switch to minimum version of armadillo supported
